### PR TITLE
test(inference): standing CI gate for composed QuaRot+Q4 forward path (#320)

### LINKED ISSUE #320: test(quarot): no CI gate covers the composed rotated+Q4 forward (f64 equivalence is pre-quant, e2e-parity is bf16)
> _Issue authored by Claude (Anthropic agent) on behalf of @ohdearquant._

## Summary

CI has two complementary equivalence defenses, but neither covers the **composed rotated+Q4 forward** end-to-end:

1. `assert_forward_equivalence_qwen35` (`quant/quarot/forward_equivalence.rs`) validates the rotation/fusion algebra **entirely in f64**, **before** quantization, at `max_abs_error ≤ 1e-5` (convert.rs:6, :390). By design it cannot see Q4 error (quant intentionally introduces 0.1–0.3 PPL, orders of magnitude above 1e-5).
2. `e2e-parity.yml` gates greedy-token agreement vs HF — but on the **bf16** model, not the Q4 artifact.

The Q4 path is component-tested (q4_weights roundtrip, Q4 GEMV/GEMM kernel tests) and the end-to-end Q4 PPL is *measured* (lattice Q4 ≈18.08, beats MLX Q4 ≈18.18) — but that PPL number is **not a CI gate**. So a regression in the interaction between QuaRot-rotated weights and the Q4 quantizer/kernels (the exact surface behind the historical "~1 PPL Q4 gap" investigation) would not fail any check; it would only show up as a manual PPL re-measure.

## Severity

Medium — defense-in-depth, not an unguarded correctness hole. The components are individually covered; what's missing is a standing composed-path regression gate. Not urgent, but valuable: a stored-golden Q4 logit gate would have shortcut the prior Q4-gap investigation (which turned out to be a stale measurement, not a bug — exactly the ambiguity a gate removes).

## Proposed work (needs design — do not land an unvalidated golden)

- Add a committed-golden regression test: rotated+quantized Qwen3.5-0.8B → first-N greedy logits/tokens, asserted against a stored golden at a **Q4-appropriate tolerance** (not 1e-5). Reference choice (MLX Q4 vs a frozen lattice-self golden) is the key design decision.
- Alternatively, extend `e2e-parity.yml` with a Q4 variant gated at token-agreement (greedy tokens are quant-robust; logit max-diff is not).
- Whichever path: the gate must use a tolerance matched to Q4 quantization error, and the golden must be regenerated deliberately (not auto-updated), mirroring the HF-parity JSON-golden pattern already in the repo.

## Provenance

Found via an automated correctness sweep of the QuaRot numerical core. The sweep also confirmed CLEAN: Hadamard construction/orthonormality + non-power-of-2 rejection, rotation axis semantics (input-side `W·Rᵀ` / output-side `R·W` with paired inverse), RMSNorm shifted-`(1+γ)` fusion on the correct axis, lm_head tied/untied handling, and Q4 scale/zero-point/rounding incl. all-zero-block degenerate-scale guard (`scale=1.0`). The only gap is the missing **composed** assertion, not any component.

### PR DESCRIPTION
**Pull request by ohdearquant**
> _PR description authored by Claude (Anthropic agent) on behalf of @ohdearquant._

## Summary

Closes #320. The QuaRot converter's own f64 forward-equivalence check
(`crates/inference/src/quant/quarot/forward_equivalence.rs`) gates the
rotation-fusion math **pre-quantization**, and `e2e-parity.yml` gates
bf16/safetensors greedy generation — but nothing exercised the runtime consumer
of a QuaRot+Q4 artifact: `MetalQwen35State::from_q4_dir` followed by greedy
decode. This PR adds a **standing regression gate** for that composed path.

## What changed

- **`crates/inference/tests/quarot_q4_composed_golden.rs`** — loads a real
  rotated Q4 artifact through the production loader (`from_q4_dir` → `generate`)
  and asserts **exact first-8 greedy token IDs** per prompt against a frozen
  lattice-self JSON golden. Tolerance is exact token agreement, **not** a logit
  tolerance: the golden is generated *from* the Q4 artifact, so Q4 quantization
  error is already baked into the expected tokens (Q4 per-weight error is orders
  of magnitude above `1e-5`). Greedy decode is deterministic and quant-robust,
  so exact token-ID equality is the honest gate.
- **`crates/inference/tests/fixtures/quarot_q4_composed_v1/…json`** — the
  committed frozen golden. Real greedy tokens from Qwen3.5-0.8B; decoded IDs are
  coherent (`short_factual` → `" Paris."`, `structured_reasoning` → correct
  `30 mph` arithmetic). **Never auto-updated.**
- **`crates/inference/src/bin/dump_quarot_q4_golden.rs`** + **`scripts/gen_quarot_q4_composed_golden.py`**
  — the regeneration binary + orchestration script. The script refuses to
  overwrite the golden without an explicit `--update-golden` (the freeze
  invariant). Regeneration procedure documented in the test module doc.
- **`.github/workflows/e2e-parity.yml`** — a new macOS `q4-composed` job (see
  "Where it runs in CI" below) + the change-detector extended to cover the new
  test / fixture / regen-script paths.
- **`crates/inference/src/bin/quantize_quarot.rs`** — `parse_u64_flex` now
  strips `_` separators so the documented seed literal `0xCAFE_BABE_DEAD_BEEF`
  actually parses (`u64::from_str_radix` rejects `_`). 3-line fix in the offline
  converter binary.

## Where it runs in CI (honest-limits — read this)

This test is **not** covered by the default workspace `cargo test` job. Under
default features and without `LATTICE_Q4_COMPOSED_GATE_ENFORCE=1`, the composed
gate **intentionally skips** (most dev machines have no ephemeral QuaRot+Q4
artifact lying around) and prints a `LATTICE_Q4_COMPOSED_SKIPPED` line.

The coverage comes from the **new `q4-composed` macOS job** in
`e2e-parity.yml`, which:
1. downloads real `Qwen/Qwen3.5-0.8B` weights (same id/cache as the existing
   `parity` job),
2. regenerates an **ephemeral** QuaRot+Q4 artifact via `quantize_quarot`,
3. runs the test with `LATTICE_Q4_COMPOSED_GATE_ENFORCE=1` +
   `LATTICE_MODEL_DIR` + `LATTICE_QUAROT_Q4_DIR` set — under enforce, a missing
   model/Q4 dir or a non-macOS/non-`metal-gpu` build **panics instead of
   skipping**, so a provisioning failure cannot masquerade as a green gate.

`q4-composed` is a **standalone, non-required** job (conditional on the
engine-change detector), reporting its own status context informationally.
It is intentionally **not** wired into `parity-gate` (the required aggregator)
yet — per ADR-066 phased-governance, new gates stay non-required until the
2026-07-15 promotion review, at which point maintainers can add `q4-composed`
alongside `embed-drift` to branch protection.

## Review round 2 (issues found → remediated)

A round-2 review found one blocker and one governance violation, both fixed
in this round:

1. **Fail-safe change detection.** The `detect-engine-changes` job used
   `if echo "$CHANGED" | grep -qE ...` under `set -euo pipefail`. `grep -q`
   exits at the first match, the upstream `echo` can then receive `SIGPIPE`,
   and under `pipefail` the pipeline reports status `141` — which the `if`
   silently treats as "no match," writing `engine=false` and skipping the
   gate as green on a large diff. Fixed with the same here-string form
   already merged on `main` for `app-binaries.yml` (`d78dc6b58`):
   `grep -E '<pattern>' <<<"$CHANGED" >/dev/null`.
2. **`q4-composed` kept non-required.** The prior round had wired
   `q4-composed` into `parity-gate`'s `needs`, added a `Q4_RESULT` check, and
   failed the required aggregator on a non-success result — contrary to the
   phased-governance shape described above. Reverted: `parity-gate`'s job body
   is now byte-identical to `main` (verified via
   `git diff origin/main -- .github/workflows/e2e-parity.yml`, which shows
   zero diff inside the `parity-gate:` job). `q4-composed` remains its own
   job, gated only on `needs.changes.outputs.engine == 'true'`.

## Mutation evidence (real, reproduced this session — replaces the prior narrative)

Ran the full PASS → FAIL → PASS cycle against the actual local Metal build,
using an ephemeral QuaRot+Q4 artifact regenerated with the exact CI seed
(`0xCAFE_BABE_DEAD_BEEF`).

**Setup**

```
target/release/quantize_quarot \
  --model-dir ~/.lattice/models/qwen3.5-0.8b \
  --output-dir <scratch>/qwen35-quarot-q4-ci \
  --seed 0xCAFE_BABE_DEAD_BEEF \
  --num-probe-tokens 4
# Forward-equiv: max_abs=7.461e-14, mean_abs=1.061e-14 (tol=1e-5)
```

**Command (all three runs, identical except fixture content)**

```
LATTICE_Q4_COMPOSED_GATE_ENFORCE=1 \
LATTICE_MODEL_DIR=~/.lattice/models/qwen3.5-0.8b \
LATTICE_QUAROT_Q4_DIR=<scratch>/qwen35-quarot-q4-ci \
cargo test --release -p lattice-inference --test quarot_q4_composed_golden \
    --features "f16,metal-gpu" -- --nocapture
```

**Fixture sha256** (`crates/inference/tests/fixtures/quarot_q4_composed_v1/qwen35_0_8b_quarot_q4_greedy_tokens.json`):
- before: `8ffafef227c9f47cff29ede519d30aa67398c72d0f5c81af5e5573f22497c603`
- mutated (first token of `short_factual` case: `11751` → `11752`):
  `fb1a63dd7cfb9d254b8551e9c66e6901e72eaab2ff0790ec9cdf33fc6f022807`
- after revert: `8ffafef227c9f47cff29ede519d30aa67398c72d0f5c81af5e5573f22497c603`
  (byte-identical to before)

```
Run 1 (clean, before any mutation):
  test quarot_q4_composed_golden_fixture_is_valid ... ok
  test quarot_q4_composed_greedy_tokens_match_lattice_self_golden ... ok
  test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out

Run 2 (fixture mutated: expected_generated_ids[0] 11751 -> 11752):
  QUAROT_Q4_COMPOSED_MISMATCH case=short_factual prompt="The capital of France is"
    expected=[11752, 13, 198, 760, 6511, 314, 9338, 369]
    actual=  [11751, 13, 198, 760, 6511, 314, 9338, 369]
  thread 'quarot_q4_composed_greedy_tokens_match_lattice_self_golden' panicked:
  assertion `left == right` failed: QuaRot Q4 composed-path greedy token
  mismatch for case 'short_factual' ...
  test result: FAILED. 1 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out

Run 3 (fixture reverted via reverse-edit, sha256 matches Run 1's fixture):
  test quarot_q4_composed_golden_fixture_is_valid ... ok
  test quarot_q4_composed_greedy_tokens_match_lattice_self_golden ... ok
  test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

**Bonus — kernel-side mutation (proves sensitivity to the runtime Q4 dequant,
not just the fixture).** In `crates/inference/src/forward/metal_qwen35.rs`,
the per-block Q4->f16 dequant loop computes `w = nibble * scale + bias`. Flipped
the sign to `w = nibble * scale - bias` (source sha256 before/after identical:
`825b01158c51b5fa43377ea67ff4f49977823567b7c74195642ef1e877549856`):

```
Run 4 (dequant bias sign flipped, rebuilt):
  QUAROT_Q4_COMPOSED_MISMATCH case=short_factual prompt="The capital of France is"
    expected=[11751, 13, 198, 760, 6511, 314, 9338, 369]
    actual=  [198, 198, 198, 198, 401, 11, 11, 11]
  test result: FAILED. 1 passed; 1 failed; 0 ignored; 0 measured; 0 filtered out

Run 5 (kernel reverted, source sha256 matches original, rebuilt):
  test quarot_q4_composed_greedy_tokens_match_lattice_self_golden ... ok
  test result: ok. 2 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out
```

Both `git diff` on the fixture and on `metal_qwen35.rs` show zero lines after
revert — the mutation-evidence testing left the working tree clean; only
`.github/workflows/e2e-parity.yml` is a real, committed change in this PR.

## bench-compare

**Not required** — this is a test + CI-wiring change. The only runtime-adjacent
edit is a 3-line seed-parsing fix in the **offline** `quantize_quarot` converter
binary (not the inference/decode hot path); no `crates/inference/src` or
`crates/embed/src` runtime forward code was touched by the committed diff (the
kernel mutation above was reverted before commit).

## Verified locally

- `cargo test --workspace` — passes (composed gate cleanly skips under default
  features).
- Enforced gate against the real regenerated artifact — 2 passed (see mutation
  evidence above for the full PASS/FAIL/PASS cycle, run this session).
- Fail-closed check: `LATTICE_Q4_COMPOSED_GATE_ENFORCE=1` with env unset →
  panics (does not skip).
- `cargo clippy --workspace -- -D warnings` clean.
- Workflow YAML parses; `parity-gate`'s `needs` is `[changes, parity,
  embed-drift]` — identical to `main`, no `q4-composed` reference.

Do **not** merge until the `q4 composed golden` CI check is green on this PR.

### CODE REVIEW

**Comment by ohdearquant** [comment]
Round-2 review remediation — reproduced the mutation-evidence cycle locally
this session (real Metal build + real ephemeral QuaRot+Q4 artifact, seed
`0xCAFE_BABE_DEAD_BEEF`, `--num-probe-tokens 4`, forward-equiv max_abs=7.461e-14).

**Fixture-level PASS -> FAIL -> PASS**

sha256 of `crates/inference/tests/fixtures/quarot_q4_composed_v1/qwen35_0_8b_quarot_q4_greedy_tokens.json`:
- before: `8ffafef227c9f47cff29ede519d30aa67398c72d0f5c81af5e5573f22497c603`
- mutated (`expected_generated_ids[0]`: `11751` -> `11752`): `fb1a63dd7cfb9d254b8551e9c66e6901e72eaab2ff0790ec9cdf33fc6f022807`
- after revert: `8ffafef227c9f47cff29ede519d30aa67398c72d0f5c81af5e5573f22497c603` (byte-identical)

```
Run 1 (clean):    test result: ok. 2 passed; 0 failed
Run 2 (mutated):  QUAROT_Q4_COMPOSED_MISMATCH case=short_factual
                  expected=[11752, 13, 198, 760, 6511, 314, 9338, 369]
                  actual=  [11751, 13, 198, 760, 6511, 314, 9338, 369]
                  test result: FAILED. 1 passed; 1 failed
Run 3 (reverted): test result: ok. 2 passed; 0 failed
```

**Bonus: kernel-level PASS -> FAIL -> PASS** (proves sensitivity to the runtime
Q4 dequant, not just the fixture). Flipped the bias sign in the per-block
dequant in `crates/inference/src/forward/metal_qwen35.rs`
(`w = nibble*scale + bias` -> `w = nibble*scale - bias`), rebuilt, ran, reverted,
rebuilt, ran again. Source sha256 before/after: `825b01158c51b5fa43377ea67ff4f49977823567b7c74195642ef1e877549856` (identical).

```
Run 4 (kernel mutated): actual=[198, 198, 198, 198, 401, 11, 11, 11] (garbage) -> FAILED
Run 5 (kernel reverted): test result: ok. 2 passed; 0 failed
```

`git diff` on both files is empty after revert — only
`.github/workflows/e2e-parity.yml` is a real change in this PR (the two
findings below).

Also fixed in this round:
1. **SIGPIPE fail-open** in `detect-engine-changes` (`echo "$CHANGED" | grep -qE ...`
   under `pipefail` -> here-string, matching the merged fix in `app-binaries.yml` @ d78dc6b58).
2. **`q4-composed` de-wired from `parity-gate`** — restored `parity-gate`'s job
   body to byte-identical with `main` (`git diff origin/main -- .github/workflows/e2e-parity.yml`
   shows zero diff inside the `parity-gate:` job). `q4-composed` stays a
   standalone, non-required, informational status check per ADR-066 phased
   governance until the 2026-07-15 promotion review.
